File: bindings/typescript/packages/aster/src/trust/gossip.ts
/**
 * Producer mesh gossip: signing, verification, message handling, heartbeat.
 *
 * Spec reference: Aster-trust-spec.md S2.3, S2.6.  Plan: ASTER_PLAN.md S14.2-14.6.
 *
 * NOT to be confused with registry/gossip.ts which handles service discovery gossip.
 * This module handles the producer mesh gossip protocol -- signing, verification,
 * message dispatch, and lease heartbeat.
 *
 * Canonical signing bytes (normative -- S2.6, ASTER_PLAN.md S14.2):
 *
 *     u8(type) || payload || sender.encode('utf-8') || u64_be(epoch_ms)
 *
 * Do NOT reorder epoch_ms before sender+payload -- the spec fixes this
 * byte order and any deviation breaks cross-implementation verification.
 *
 * Topic derivation (S2.3):
 *
 *     blake3(root_pubkey + "aster-producer-mesh" + salt).digest().slice(0, 32)
 */

import { sign, verify, concatBytes, bytesToHex } from './credentials.js';
import { validateRcan } from './rcan.js';
import type { MeshState } from './mesh.js';
import type { ClockDriftTracker } from './clock.js';

// ── Types ───────────────────────────────────────────────────────────────────

export interface ProducerMessage {
  type: number;
  payload: Uint8Array;
  sender: string;       // endpoint ID hex
  epochMs: number;
  signature: Uint8Array; // 64 bytes
}

export enum ProducerMessageType {
  INTRODUCE = 1,
  DEPART = 2,
  CONTRACT_PUBLISHED = 3,
  LEASE_UPDATE = 4,
}

export interface ContractPublishedPayload {
  serviceName: string;
  version: number;
  contractCollectionHash: string;
}

export interface LeaseUpdatePayload {
  serviceName: string;
  version: number;
  contractId: string;
  healthStatus: string;
  addressingInfo: Record<string, string>;
}

export interface HandleMessageOptions {
  state: MeshState;
  config: { replayWindowMs: number };
  peerPubkeys: Map<string, Uint8Array>;  // endpoint_id -> pubkey
  registryCallback?: (eventType: string, payload: ContractPublishedPayload | LeaseUpdatePayload) => void;
  driftTracker?: ClockDriftTracker;
  onSelfDeparture?: () => void;
  logger?: { info(...args: unknown[]): void; warn(...args: unknown[]): void; debug(...args: unknown[]): void; error(...args: unknown[]): void };
}

/** Maximum gossip payload size (matches Python aster.limits.MAX_GOSSIP_PAYLOAD_SIZE). */
const MAX_GOSSIP_PAYLOAD_SIZE = 64 * 1024;

// ── Topic derivation ────────────────────────────────────────────────────────

/**
 * Derive the 32-byte gossip topic for the producer mesh.
 *
 * blake3(root_pubkey + "aster-producer-mesh" + salt).slice(0, 32)
 *
 * Falls back to sha256 if blake3 is not available.
 */
export async function deriveGossipTopic(rootPubkey: Uint8Array, salt: Uint8Array): Promise<Uint8Array> {
  const label = new TextEncoder().encode('aster-producer-mesh');
  const data = concatBytes([rootPubkey, label, salt]);

  try {
    // Try @noble/hashes blake3 first
    // @ts-ignore — optional dependency
    const { blake3 } = await import('@noble/hashes/blake3') as any;
    return blake3(data, { dkLen: 32 });
  } catch {
    // Fallback: sha256 via node:crypto (note: Python uses blake3)
    const { createHash } = await import('node:crypto');
    const hash = createHash('sha256').update(data).digest();
    return new Uint8Array(hash.buffer, hash.byteOffset, 32);
  }
}

// ── Canonical signing bytes ─────────────────────────────────────────────────

/**
 * Return the canonical bytes that are signed / verified.
 *
 * Normative byte order (S2.6):
 *     u8(type) || payload || sender.encode('utf-8') || u64_be(epoch_ms)
 */
export function producerMessageSigningBytes(
  msgType: number,
  payload: Uint8Array,
  sender: string,
  epochMs: number,
): Uint8Array {
  const typeByte = new Uint8Array([msgType & 0xff]);
  const senderBytes = new TextEncoder().encode(sender);
  const epochBytes = writeU64BE(epochMs);
  return concatBytes([typeByte, payload, senderBytes, epochBytes]);
}

/** u64 big-endian encoding. */
function writeU64BE(value: number): Uint8Array {
  const buf = new ArrayBuffer(8);
  const view = new DataView(buf);
  view.setUint32(0, Math.floor(value / 0x100000000));
  view.setUint32(4, value >>> 0);
  return new Uint8Array(buf);
}

// ── Sign / verify ───────────────────────────────────────────────────────────

/**
 * Create and sign a ProducerMessage.
 *
 * @param msgType        ProducerMessageType integer value.
 * @param payload        Serialized per-type payload bytes.
 * @param sender         This node's endpoint_id (hex string).
 * @param epochMs        Wall-clock timestamp in milliseconds.
 * @param signingKeyRaw  32-byte raw ed25519 private key seed.
 * @returns A fully signed ProducerMessage.
 */
export async function signProducerMessage(
  msgType: number,
  payload: Uint8Array,
  sender: string,
  epochMs: number,
  signingKeyRaw: Uint8Array,
): Promise<ProducerMessage> {
  const toSign = producerMessageSigningBytes(msgType, payload, sender, epochMs);
  const signature = await sign(signingKeyRaw, toSign);
  return {
    type: msgType,
    payload,
    sender,
    epochMs,
    signature,
  };
}

/**
 * Verify a ProducerMessage's signature against the sender's public key.
 *
 * @returns true on success, false on any verification failure.
 */
export async function verifyProducerMessage(
  msg: ProducerMessage,
  peerPubkeyRaw: Uint8Array,
): Promise<boolean> {
  try {
    const toVerify = producerMessageSigningBytes(msg.type, msg.payload, msg.sender, msg.epochMs);
    return await verify(peerPubkeyRaw, toVerify, msg.signature);
  } catch {
    return false;
  }
}

// ── Gossip handler ──────────────────────────────────────────────────────────

/**
 * Process one inbound ProducerMessage.
 *
 * Normative processing order (S2.6, S2.10):
 * 1. Replay-window check.
 * 2. Sender membership check.
 * 3. Signature verification.
 * 4. Track clock offset / run drift detection.
 * 5. Dispatch by message type.
 *
 * @returns true if the message was accepted and dispatched, false otherwise.
 */
export async function handleProducerMessage(
  msg: ProducerMessage,
  opts: HandleMessageOptions,
): Promise<boolean> {
  const { state, config, peerPubkeys, registryCallback, driftTracker, onSelfDeparture, logger } = opts;
  const nowMs = Date.now();

  // 1. Replay-window check
  const delta = Math.abs(nowMs - msg.epochMs);
  if (delta > config.replayWindowMs) {
    logger?.debug(
      'gossip: dropping message from %s (outside replay window: delta=%dms)',
      msg.sender,
      delta,
    );
    return false;
  }

  // 2. Sender membership check
  if (!state.isPeerAccepted(msg.sender)) {
    logger?.warn(
      'gossip: SECURITY ALERT -- message from non-accepted sender %s; ' +
      'possible salt leak or deauthorized node still subscribed',
      msg.sender,
    );
    return false;
  }

  // 3. Signature verification
  const peerPubkey = peerPubkeys.get(msg.sender);
  if (peerPubkey == null) {
    logger?.warn(
      'gossip: SECURITY ALERT -- no public key for accepted sender %s; dropping message',
      msg.sender,
    );
    return false;
  }

  if (!(await verifyProducerMessage(msg, peerPubkey))) {
    logger?.warn(
      'gossip: SECURITY ALERT -- invalid signature from accepted sender %s',
      msg.sender,
    );
    return false;
  }

  // 4. Clock offset tracking + drift detection
  if (driftTracker != null) {
    const newlyIsolated = driftTracker.update(msg.sender, msg.epochMs);
    if (newlyIsolated) {
      logger?.warn(
        'gossip: peer %s clock drift exceeds tolerance; isolating',
        msg.sender,
      );
    } else if (!driftTracker.isIsolated(msg.sender)) {
      // Peer recovered -- already handled by driftTracker.update
    }
  }

  // 5. Message dispatch
  switch (msg.type) {
    case ProducerMessageType.INTRODUCE:
      _handleIntroduce(msg, state, logger);
      break;
    case ProducerMessageType.DEPART:
      _handleDepart(msg, state, driftTracker, onSelfDeparture, logger);
      break;
    case ProducerMessageType.CONTRACT_PUBLISHED:
      _handleContractPublished(msg, state, driftTracker, registryCallback, logger);
      break;
    case ProducerMessageType.LEASE_UPDATE:
      _handleLeaseUpdate(msg, state, driftTracker, registryCallback, logger);
      break;
    default:
      logger?.debug('gossip: unknown message type %d from %s; dropping', msg.type, msg.sender);
      return false;
  }

  return true;
}

// ── Type-specific dispatch helpers ──────────────────────────────────────────

function _handleIntroduce(
  msg: ProducerMessage,
  state: MeshState,
  logger?: HandleMessageOptions['logger'],
): void {
  const [ok, reason] = validateRcan(msg.payload);
  if (!ok) {
    logger?.debug('gossip: Introduce from %s has invalid rcan: %s', msg.sender, reason);
    return;
  }
  state.addPeer(msg.sender);
  logger?.info('gossip: Introduce accepted -- %s joined the mesh', msg.sender);
}

function _handleDepart(
  msg: ProducerMessage,
  state: MeshState,
  driftTracker?: ClockDriftTracker,
  onSelfDeparture?: () => void,
  logger?: HandleMessageOptions['logger'],
): void {
  state.remove(msg.sender);
  driftTracker?.removePeer(msg.sender);
  logger?.info('gossip: Depart -- %s left the mesh', msg.sender);
  void onSelfDeparture; // used by caller for self-departure detection
}

function _handleContractPublished(
  msg: ProducerMessage,
  _state: MeshState,
  driftTracker?: ClockDriftTracker,
  registryCallback?: HandleMessageOptions['registryCallback'],
  logger?: HandleMessageOptions['logger'],
): void {
  if (driftTracker?.isIsolated(msg.sender)) {
    logger?.debug('gossip: ContractPublished from drift-isolated peer %s; skipping', msg.sender);
    return;
  }
  if (registryCallback != null) {
    try {
      if (msg.payload.length > MAX_GOSSIP_PAYLOAD_SIZE) {
        logger?.warn('gossip: payload too large (%d bytes), dropping', msg.payload.length);
        return;
      }
      const d = JSON.parse(new TextDecoder().decode(msg.payload));
      const payload: ContractPublishedPayload = {
        serviceName: d.service_name,
        version: Number(d.version),
        contractCollectionHash: d.contract_collection_hash,
      };
      registryCallback('contract_published', payload);
    } catch (exc) {
      logger?.debug('gossip: malformed ContractPublished payload: %s', exc);
    }
  }
}

function _handleLeaseUpdate(
  msg: ProducerMessage,
  _state: MeshState,
  driftTracker?: ClockDriftTracker,
  registryCallback?: HandleMessageOptions['registryCallback'],
  logger?: HandleMessageOptions['logger'],
): void {
  if (driftTracker?.isIsolated(msg.sender)) {
    logger?.debug('gossip: LeaseUpdate from drift-isolated peer %s; skipping', msg.sender);
    return;
  }
  if (registryCallback != null) {
    try {
      if (msg.payload.length > MAX_GOSSIP_PAYLOAD_SIZE) {
        logger?.warn('gossip: LeaseUpdate payload too large (%d bytes)', msg.payload.length);
        return;
      }
      const d = JSON.parse(new TextDecoder().decode(msg.payload));
      const payload: LeaseUpdatePayload = {
        serviceName: d.service_name,
        version: Number(d.version),
        contractId: d.contract_id,
        healthStatus: d.health_status,
        addressingInfo: d.addressing_info ?? {},
      };
      registryCallback('lease_update', payload);
    } catch (exc) {
      logger?.debug('gossip: malformed LeaseUpdate payload: %s', exc);
    }
  }
}

// ── Payload serializers (JSON-based for Phase 12) ───────────────────────────

/** Encode IntroducePayload. The rcan grant is stored as raw bytes. */
export function encodeIntroducePayload(rcan: Uint8Array): Uint8Array {
  return rcan;
}

/** Encode DepartPayload as UTF-8 JSON. */
export function encodeDepartPayload(reason = ''): Uint8Array {
  return new TextEncoder().encode(JSON.stringify({ reason }));
}

/** Encode ContractPublishedPayload as UTF-8 JSON. */
export function encodeContractPublishedPayload(
  serviceName: string,
  version: number,
  contractCollectionHash: string,
): Uint8Array {
  return new TextEncoder().encode(
    JSON.stringify({
      service_name: serviceName,
      version,
      contract_collection_hash: contractCollectionHash,
    }),
  );
}

/** Encode LeaseUpdatePayload as UTF-8 JSON. */
export function encodeLeaseUpdatePayload(
  serviceName: string,
  version: number,
  contractId: string,
  healthStatus: string,
  addressingInfo: Record<string, string> = {},
): Uint8Array {
  return new TextEncoder().encode(
    JSON.stringify({
      service_name: serviceName,
      version,
      contract_id: contractId,
      health_status: healthStatus,
      addressing_info: addressingInfo,
    }),
  );
}

// ── Lease heartbeat ─────────────────────────────────────────────────────────

export interface StartLeaseHeartbeatOptions {
  /** A gossip topic handle with a broadcast(data: Uint8Array) method. */
  gossipTopicHandle: { broadcast(data: Uint8Array): Promise<void> };
  sender: string;
  signingKeyRaw: Uint8Array;
  serviceName: string;
  version: number;
  contractId: string;
  healthGetter: () => string;
  heartbeatIntervalMs?: number;
  addressingInfo?: Record<string, string>;
  logger?: HandleMessageOptions['logger'];
}

/**
 * Sign and broadcast a single LEASE_UPDATE message.
 */
export async function runLeaseHeartbeat(
  gossipTopicHandle: { broadcast(data: Uint8Array): Promise<void> },
  sender: string,
  signingKeyRaw: Uint8Array,
  serviceName: string,
  version: number,
  contractId: string,
  healthGetter: () => string,
  addressingInfo: Record<string, string> = {},
  logger?: HandleMessageOptions['logger'],
): Promise<void> {
  const epochMs = Date.now();
  const payload = encodeLeaseUpdatePayload(
    serviceName,
    version,
    contractId,
    healthGetter(),
    addressingInfo,
  );
  const msg = await signProducerMessage(
    ProducerMessageType.LEASE_UPDATE,
    payload,
    sender,
    epochMs,
    signingKeyRaw,
  );

  // Serialize the signed envelope as JSON for the gossip wire.
  const wire = new TextEncoder().encode(
    JSON.stringify({
      type: msg.type,
      payload: bytesToHex(msg.payload),
      sender: msg.sender,
      epoch_ms: msg.epochMs,
      signature: bytesToHex(msg.signature),
    }),
  );

  try {
    await gossipTopicHandle.broadcast(wire);
    logger?.debug(
      'heartbeat: LeaseUpdate broadcast for %s v%d (epoch=%d)',
      serviceName,
      version,
      epochMs,
    );
  } catch (exc) {
    logger?.warn('heartbeat: broadcast failed: %s', exc);
  }
}

/**
 * Spawn a periodic lease heartbeat that broadcasts LEASE_UPDATE messages.
 *
 * @returns An object with a stop() method to cancel the heartbeat.
 */
export function startLeaseHeartbeat(opts: StartLeaseHeartbeatOptions): { stop: () => void } {
  const {
    gossipTopicHandle,
    sender,
    signingKeyRaw,
    serviceName,
    version,
    contractId,
    healthGetter,
    heartbeatIntervalMs = 900_000,
    addressingInfo = {},
    logger,
  } = opts;

  const timer = setInterval(() => {
    runLeaseHeartbeat(
      gossipTopicHandle,
      sender,
      signingKeyRaw,
      serviceName,
      version,
      contractId,
      healthGetter,
      addressingInfo,
      logger,
    ).catch((err) => {
      logger?.warn('heartbeat: tick failed: %s', err);
    });
  }, heartbeatIntervalMs);

  return {
    stop() {
      clearInterval(timer);
    },
  };
}

File: bindings/python/aster/trust/gossip.py
"""
aster.trust.gossip — Producer mesh gossip: signing, verification, and dispatch.

Spec reference: Aster-trust-spec.md §2.3, §2.6.  Plan: ASTER_PLAN.md §14.2–14.6.

Canonical signing bytes (normative — §2.6, ASTER_PLAN.md §14.2):

    u8(type) || payload || sender.encode('utf-8') || u64_be(epoch_ms)

Do NOT reorder ``epoch_ms`` before ``sender+payload`` — the spec fixes this
byte order and any deviation breaks cross-implementation verification.

Topic derivation (§2.3):

    blake3(root_pubkey + b"aster-producer-mesh" + salt).digest()[:32]
"""

from __future__ import annotations

import logging
import struct
import time
from typing import Any, Callable

from .mesh import (
    ClockDriftConfig,
    ContractPublishedPayload,
    LeaseUpdatePayload,
    MeshState,
    ProducerMessage,
    ProducerMessageType,
)
from .rcan import validate_rcan

logger = logging.getLogger(__name__)


# ── Topic derivation ──────────────────────────────────────────────────────────


def derive_gossip_topic(root_pubkey: bytes, salt: bytes) -> bytes:
    """Derive the 32-byte gossip topic for the producer mesh.

    ``blake3(root_pubkey + b"aster-producer-mesh" + salt).digest()``

    The salt keeps the topic private to admitted producers.
    """
    import blake3  # type: ignore[import]

    data = root_pubkey + b"aster-producer-mesh" + salt
    return blake3.blake3(data).digest()


# ── Canonical signing bytes ───────────────────────────────────────────────────


def producer_message_signing_bytes(
    msg_type: int,
    payload: bytes,
    sender: str,
    epoch_ms: int,
) -> bytes:
    """Return the canonical bytes that are signed / verified.

    Normative byte order (§2.6):
        u8(type) || payload || sender.encode('utf-8') || u64_be(epoch_ms)
    """
    return b"".join(
        [
            struct.pack("B", msg_type),
            payload,
            sender.encode("utf-8"),
            struct.pack(">Q", epoch_ms),
        ]
    )


# ── Sign / verify ─────────────────────────────────────────────────────────────


def sign_producer_message(
    msg_type: int,
    payload: bytes,
    sender: str,
    epoch_ms: int,
    signing_key_raw: bytes,
) -> ProducerMessage:
    """Create and sign a ProducerMessage.

    Args:
        msg_type:        ProducerMessageType integer value.
        payload:         Serialized per-type payload bytes.
        sender:          This node's endpoint_id (hex string).
        epoch_ms:        Wall-clock timestamp in milliseconds.
        signing_key_raw: 32-byte raw ed25519 private key seed.

    Returns:
        A fully signed ``ProducerMessage``.
    """
    from .signing import load_private_key

    to_sign = producer_message_signing_bytes(msg_type, payload, sender, epoch_ms)
    privkey = load_private_key(signing_key_raw)
    signature = privkey.sign(to_sign)
    return ProducerMessage(
        type=msg_type,
        payload=payload,
        sender=sender,
        epoch_ms=epoch_ms,
        signature=signature,
    )


def verify_producer_message(msg: ProducerMessage, peer_pubkey_raw: bytes) -> bool:
    """Verify a ProducerMessage's signature against the sender's public key.

    Returns True on success, False on any verification failure.
    """
    from cryptography.exceptions import InvalidSignature

    from .signing import load_public_key

    try:
        to_verify = producer_message_signing_bytes(
            msg.type, msg.payload, msg.sender, msg.epoch_ms
        )
        pubkey = load_public_key(peer_pubkey_raw)
        pubkey.verify(msg.signature, to_verify)
        return True
    except (InvalidSignature, Exception):  # noqa: BLE001
        return False


# ── Gossip handler ────────────────────────────────────────────────────────────


async def handle_producer_message(
    msg: ProducerMessage,
    state: MeshState,
    config: ClockDriftConfig,
    peer_pubkeys: dict[str, bytes],
    registry_callback: Callable[[str, object], None] | None = None,
    drift_detector: "ClockDriftDetector | None" = None,  # noqa: F821
    on_self_departure: Callable[[], None] | None = None,
) -> None:
    """Process one inbound ProducerMessage.

    Normative processing order (§2.6, §2.10):
    1. Replay-window check.
    2. Sender membership check.
    3. Signature verification.
    4. Track clock offset / run drift detection.
    5. Dispatch by message type.

    Args:
        msg:               Incoming message.
        state:             Mutable MeshState shared by the node.
        config:            Clock drift / replay-window configuration.
        peer_pubkeys:      Maps endpoint_id → raw 32-byte ed25519 public key.
        registry_callback: Optional callback(event_type, payload) for Phase 10
                           integration (ContractPublished / LeaseUpdate).
        drift_detector:    Optional ClockDriftDetector instance.
        on_self_departure: Optional zero-arg callback invoked when self-departure
                           is triggered.
    """
    now_ms = int(time.time() * 1000)

    # ── 1. Replay-window check ────────────────────────────────────────────────
    delta = abs(now_ms - msg.epoch_ms)
    if delta > config.replay_window_ms:
        logger.debug(
            "gossip: dropping message from %s (outside replay window: delta=%dms)",
            msg.sender,
            delta,
        )
        return

    # ── 2. Sender membership check ────────────────────────────────────────────
    if msg.sender not in state.accepted_producers:
        logger.warning(
            "gossip: SECURITY ALERT — message from non-accepted sender %s; "
            "possible salt leak or deauthorized node still subscribed",
            msg.sender,
        )
        return

    # ── 3. Signature verification ─────────────────────────────────────────────
    peer_pubkey = peer_pubkeys.get(msg.sender)
    if peer_pubkey is None:
        # Cannot verify — no public key on file for this sender.
        logger.warning(
            "gossip: SECURITY ALERT — no public key for accepted sender %s; "
            "dropping message",
            msg.sender,
        )
        return

    if not verify_producer_message(msg, peer_pubkey):
        logger.warning(
            "gossip: SECURITY ALERT — invalid signature from accepted sender %s",
            msg.sender,
        )
        return

    # ── 4. Clock offset tracking + drift detection ────────────────────────────
    offset = now_ms - msg.epoch_ms
    state.peer_offsets[msg.sender] = offset

    if drift_detector is not None:
        drift_detector.track_offset(msg.sender, msg.epoch_ms)

        # Peer drift isolation / recovery
        if drift_detector.peer_in_drift(msg.sender):
            if msg.sender not in state.drift_isolated:
                logger.warning(
                    "gossip: peer %s clock drift exceeds tolerance (offset=%dms); "
                    "isolating",
                    msg.sender,
                    offset,
                )
                state.drift_isolated.add(msg.sender)
        else:
            # Fresh acceptable message — recover from isolation if present
            if msg.sender in state.drift_isolated:
                logger.info(
                    "gossip: peer %s recovered from drift isolation",
                    msg.sender,
                )
                state.drift_isolated.discard(msg.sender)

        # Self-drift check (self-offset ≈ 0 if our clock is accurate)
        self_offset_estimate = 0  # We just sent, so our offset is near zero
        if drift_detector.self_in_drift(self_offset_estimate) and not state.mesh_dead:
            logger.error(
                "gossip: self-departure triggered — local clock deviates from "
                "mesh median; broadcasting Depart and suppressing further sends"
            )
            state.mesh_dead = True
            if on_self_departure is not None:
                on_self_departure()
            return

    # ── 5. Message dispatch ───────────────────────────────────────────────────
    msg_type = msg.type
    if msg_type == ProducerMessageType.INTRODUCE:
        _handle_introduce(msg, state)
    elif msg_type == ProducerMessageType.DEPART:
        _handle_depart(msg, state, drift_detector)
    elif msg_type == ProducerMessageType.CONTRACT_PUBLISHED:
        _handle_contract_published(msg, state, registry_callback)
    elif msg_type == ProducerMessageType.LEASE_UPDATE:
        _handle_lease_update(msg, state, registry_callback)
    else:
        logger.debug("gossip: unknown message type %d from %s; dropping", msg_type, msg.sender)


# ── Type-specific dispatch helpers ────────────────────────────────────────────


def _handle_introduce(msg: ProducerMessage, state: MeshState) -> None:
    ok, reason = validate_rcan(msg.payload)
    if not ok:
        logger.debug(
            "gossip: Introduce from %s has invalid rcan: %s", msg.sender, reason
        )
        return
    state.accepted_producers.add(msg.sender)
    logger.info("gossip: Introduce accepted — %s joined the mesh", msg.sender)


def _handle_depart(
    msg: ProducerMessage,
    state: MeshState,
    drift_detector: "ClockDriftDetector | None",  # noqa: F821
) -> None:
    state.accepted_producers.discard(msg.sender)
    state.drift_isolated.discard(msg.sender)
    state.peer_offsets.pop(msg.sender, None)
    if drift_detector is not None:
        drift_detector.remove_peer(msg.sender)
    logger.info("gossip: Depart — %s left the mesh", msg.sender)


def _handle_contract_published(
    msg: ProducerMessage,
    state: MeshState,
    registry_callback: Callable[[str, object], None] | None,
) -> None:
    if msg.sender in state.drift_isolated:
        logger.debug(
            "gossip: ContractPublished from drift-isolated peer %s; skipping",
            msg.sender,
        )
        return
    if registry_callback is not None:
        payload = ContractPublishedPayload.__new__(ContractPublishedPayload)
        # Parse payload bytes as simple JSON to avoid heavy Fory dependency here.
        import json

        try:
            from aster.limits import MAX_GOSSIP_PAYLOAD_SIZE
            raw = msg.payload
            if len(raw) > MAX_GOSSIP_PAYLOAD_SIZE:
                logger.warning("gossip: payload too large (%d bytes), dropping", len(raw))
                return
            d = json.loads(raw.decode("utf-8"))
            payload.service_name = d["service_name"]
            payload.version = int(d["version"])
            payload.contract_collection_hash = d["contract_collection_hash"]
            registry_callback("contract_published", payload)
        except Exception as exc:  # noqa: BLE001
            logger.debug("gossip: malformed ContractPublished payload: %s", exc)


def _handle_lease_update(
    msg: ProducerMessage,
    state: MeshState,
    registry_callback: Callable[[str, object], None] | None,
) -> None:
    if msg.sender in state.drift_isolated:
        logger.debug(
            "gossip: LeaseUpdate from drift-isolated peer %s; skipping",
            msg.sender,
        )
        return
    if registry_callback is not None:
        import json

        try:
            from aster.limits import MAX_GOSSIP_PAYLOAD_SIZE
            if len(msg.payload) > MAX_GOSSIP_PAYLOAD_SIZE:
                logger.warning("gossip: LeaseUpdate payload too large (%d bytes)", len(msg.payload))
                return
            d = json.loads(msg.payload.decode("utf-8"))
            payload = LeaseUpdatePayload(
                service_name=d["service_name"],
                version=int(d["version"]),
                contract_id=d["contract_id"],
                health_status=d["health_status"],
                addressing_info=d.get("addressing_info", {}),
            )
            registry_callback("lease_update", payload)
        except Exception as exc:  # noqa: BLE001
            logger.debug("gossip: malformed LeaseUpdate payload: %s", exc)


# ── Lease heartbeat ──────────────────────────────────────────────────────────


async def run_lease_heartbeat(
    gossip_topic_handle: object,
    sender: str,
    signing_key_raw: bytes,
    service_name: str,
    version: int,
    contract_id: str,
    health_getter: "Callable[[], str]",
    heartbeat_interval_ms: int = 900_000,
    addressing_info: "dict[str, str] | None" = None,
) -> None:
    """Broadcast a signed LeaseUpdate message every ``heartbeat_interval_ms``.

    This coroutine runs until cancelled.  Callers should wrap it with
    ``asyncio.create_task`` and cancel the task on shutdown.

    Args:
        gossip_topic_handle: A ``GossipTopicHandle`` with a
            ``broadcast(data: bytes) -> Coroutine`` method.
        sender:              This node's endpoint_id (hex string).
        signing_key_raw:     32-byte raw ed25519 private key seed.
        service_name:        Service name to include in the LeaseUpdate payload.
        version:             Service version.
        contract_id:         Contract ID (64-char hex).
        health_getter:       Zero-arg callable returning current health status
                             string.  Called fresh on each heartbeat so state
                             transitions are reflected immediately.
        heartbeat_interval_ms: Broadcast cadence in milliseconds (default 15 min).
        addressing_info:     Optional addressing metadata forwarded in the payload.
    """
    import asyncio

    interval_s = heartbeat_interval_ms / 1000.0
    try:
        while True:
            await asyncio.sleep(interval_s)
            epoch_ms = int(time.time() * 1000)
            payload = encode_lease_update_payload(
                service_name=service_name,
                version=version,
                contract_id=contract_id,
                health_status=health_getter(),
                addressing_info=addressing_info,
            )
            msg = sign_producer_message(
                msg_type=ProducerMessageType.LEASE_UPDATE,
                payload=payload,
                sender=sender,
                epoch_ms=epoch_ms,
                signing_key_raw=signing_key_raw,
            )
            # Serialize the signed envelope as JSON for the gossip wire.
            import json as _json
            wire = _json.dumps(
                {
                    "type": msg.type,
                    "payload": msg.payload.hex(),
                    "sender": msg.sender,
                    "epoch_ms": msg.epoch_ms,
                    "signature": msg.signature.hex(),
                },
                separators=(",", ":"),
            ).encode("utf-8")
            try:
                await gossip_topic_handle.broadcast(wire)
                logger.debug(
                    "heartbeat: LeaseUpdate broadcast for %s v%d (epoch=%d)",
                    service_name,
                    version,
                    epoch_ms,
                )
            except Exception as exc:  # noqa: BLE001
                logger.warning("heartbeat: broadcast failed: %s", exc)
    except asyncio.CancelledError:
        pass


def start_lease_heartbeat(
    gossip_topic_handle: object,
    sender: str,
    signing_key_raw: bytes,
    service_name: str,
    version: int,
    contract_id: str,
    health_getter: "Callable[[], str]",
    heartbeat_interval_ms: int = 900_000,
    addressing_info: "dict[str, str] | None" = None,
) -> "Any":
    """Spawn ``run_lease_heartbeat`` as an asyncio background task.

    Returns the ``asyncio.Task`` so the caller can cancel it on shutdown.
    """
    import asyncio

    return asyncio.create_task(
        run_lease_heartbeat(
            gossip_topic_handle=gossip_topic_handle,
            sender=sender,
            signing_key_raw=signing_key_raw,
            service_name=service_name,
            version=version,
            contract_id=contract_id,
            health_getter=health_getter,
            heartbeat_interval_ms=heartbeat_interval_ms,
            addressing_info=addressing_info,
        ),
        name=f"lease-heartbeat-{service_name}-v{version}",
    )


# ── Payload serializers (JSON-based for Phase 12) ─────────────────────────────


def encode_introduce_payload(rcan: bytes) -> bytes:
    """Encode IntroducePayload.  The rcan grant is stored as raw bytes."""
    # Use rcan bytes directly as payload — opaque pass-through.
    return rcan


def encode_depart_payload(reason: str = "") -> bytes:
    """Encode DepartPayload as UTF-8 JSON."""
    import json

    return json.dumps({"reason": reason}, separators=(",", ":")).encode("utf-8")


def encode_contract_published_payload(
    service_name: str, version: int, contract_collection_hash: str
) -> bytes:
    """Encode ContractPublishedPayload as UTF-8 JSON."""
    import json

    return json.dumps(
        {
            "service_name": service_name,
            "version": version,
            "contract_collection_hash": contract_collection_hash,
        },
        separators=(",", ":"),
    ).encode("utf-8")


def encode_lease_update_payload(
    service_name: str,
    version: int,
    contract_id: str,
    health_status: str,
    addressing_info: dict[str, str] | None = None,
) -> bytes:
    """Encode LeaseUpdatePayload as UTF-8 JSON."""
    import json

    return json.dumps(
        {
            "service_name": service_name,
            "version": version,
            "contract_id": contract_id,
            "health_status": health_status,
            "addressing_info": addressing_info or {},
        },
        separators=(",", ":"),
    ).encode("utf-8")
